src/stream/src/executor/managed_state/join/mod.rs

```diff
@@ -326,48 +375,113 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
-            let row = row?.into_owned();
-            let pk = row.by_indices(&self.pk_indices);
 
-            entry_state.insert(pk, JoinRow::from_row(row).encode()?);
-        }
+        if self.need_degree_table {
+            let degree_table_iter_fut = self
+                .degree_state
+                .table
+                .iter_with_pk_prefix(&key, self.current_epoch);
+
+            let (table_iter, degree_table_iter) =
+                try_join(table_iter_fut, degree_table_iter_fut).await?;
+
+            // We need this because ttl may remove some entries from table but leave the entries
+            // with the same stream key in degree table.
+            let zipped_iter = zip_by_order_key(
+                table_iter,
+                &self.state.pk_indices,
+                degree_table_iter,
+                &self.degree_state.pk_indices,
+            );
+
+            #[for_await]
+            for row_and_degree in zipped_iter {
+                let (row, degree) = row_and_degree?;
+                debug_assert_eq!(degree.size(), self.degree_state.order_key_indices.len() + 1);
+                let pk = row.by_indices(&self.state.pk_indices);
+                let degree_i64 = degree
+                    .0
+                    .last()
+                    .cloned()
+                    .context("Empty row")?
+                    .context("Fail to fetch a degree")?;
+                entry_state.insert(
+                    pk,
+                    JoinRow::new(row.into_owned(), *degree_i64.as_int64() as u64).encode()?,
+                );
+            }
+        } else {
+            let table_iter = table_iter_fut.await?;
+
+            #[for_await]
+            for row in table_iter {
+                let row = row?;
+                let pk = row.by_indices(&self.state.pk_indices);
+                entry_state.insert(pk, JoinRow::new(row.into_owned(), 0).encode()?);
+            }
+        };
+
         Ok(entry_state)
     }
 
     pub async fn flush(&mut self) -> StreamExecutorResult<()> {
         self.metrics.flush();
-        self.state_table.commit(self.current_epoch).await?;
+        self.state.table.commit(self.current_epoch).await?;
+        self.degree_state.table.commit(self.current_epoch).await?;
         Ok(())
     }
 
-    /// Insert a key
+    /// Insert a join row
     pub fn insert(&mut self, key: &K, pk: Row, value: JoinRow) -> StreamExecutorResult<()> {
         if let Some(entry) = self.inner.get_mut(key) {
             entry.insert(pk, value.encode()?);
         }
-
         // If no cache maintained, only update the flush buffer.
-        self.state_table.insert(value.into_row())?;
+        let (row, degree) = value.into_table_rows(&self.state.order_key_indices);
+        self.state.table.insert(row)?;
+        self.degree_state.table.insert(degree)?;
+        Ok(())
+    }
+
+    /// Insert a row
+    pub fn insert_row(&mut self, key: &K, pk: Row, value: Row) -> StreamExecutorResult<()> {
```

Comment:
Could you please show the difference and use cases for these two `insert` methods in docs?